quic/s2n-quic-qns/src/xdp.rs

```diff
@@ -315,4 +325,64 @@ impl Xdp {
 
         Ok(provider)
     }
+
+    pub fn client(&self, addr: SocketAddr) -> Result<impl io::Provider> {
+        let udp_socket = self.udp_socket(addr)?;
+        // query the actual port that was selected for the socket in case it was `0`
+        let addr = udp_socket.local_addr()?;
+
+        let recv_udp_socket = udp_socket.try_clone();
+
+        let (umem, rx, rx_fds, tx) = self.setup()?;
+
+        self.bpf_task(addr.port(), rx_fds)?;
+
+        let io_rx = xdp_io::rx::Rx::new(rx, umem.clone());
+
+        let io_tx = {
+            let tx = xdp_io::tx::Tx::new(tx, umem, self.tx_encoder());
+
+            let udp_tx = {
+                let (udp_tx, udp_task) = tx::channel(udp_socket);
+
+                tokio::spawn(udp_task);
+
+                udp_tx.with_handle_map(|handle: &io::xdp::PathHandle| {
+                    // convert the XDP handle into the regular UDP handle
+                    handle.into()
+                })
+            };
+
+            // route any initial packets to the UDP socket so we can offload address resolution on
+            // the OS
+            tx.with_router(xdp_io::router::Router::default(), udp_tx)
+        };
+
+        let provider = Provider::builder()
+            .with_rx(io_rx)
+            .with_tx(io_tx)
+            .with_frame_size(self.frame_size as _)?
+            .build();
+
+        if let Ok(udp_socket) = recv_udp_socket {
+            let udp_socket = tokio::net::UdpSocket::from_std(udp_socket)?;
+            // Set up a task to read from the bound UDP socket
```

Comment:
So this exists just to verify everything is working properly? Would it make sense to do something more severe than logging if a packet is received?

Reply:
Yeah that's correct. I can shut down the process with an error message if you think that's better.

Another option is to merge the queues and still process them in the endpoint. But the problem with that is it'll be using normal UDP sockets and you won't really have any insight that that's happening...

Reply:
I don't really have a sense of why this would happen or how likely it is. If you think it would only happen as soon as the process starts up then maybe its better to shut down with an error, otherwise I could see this spamming logs pretty heavily.

Reply:
Yeah if this starts logging the cause would likely be that the XDP program was unloaded, which we wouldn't recover from in the current code since we'd have to re-register all of the map values. So in this example code I think it's best to shut it down.